File: 109-convert-sorted-list-to-binary-search-tree/leochang/109.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	TreeNode* sortedListToBST(ListNode* head) {
		if (head == NULL) {
			return NULL;
		}
		return recursivetoBST(head, NULL);
	}
private:
	TreeNode* recursivetoBST(ListNode *head, ListNode *tail){
		if (head == tail) {
			return NULL;
		}
		ListNode *slow = head, *fast = head;
		while (fast != tail && fast->next != tail) {
			slow = slow->next;
			fast = fast->next->next;
		}
		TreeNode *root = new TreeNode(slow->val);
		root->left = recursivetoBST(head, slow);
		root->right = recursivetoBST(slow->next, tail);
		return root;
	}
};

File: 109-convert-sorted-list-to-binary-search-tree/ken/109.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
	def sortedListToBST(self, head):
		"""
		:type head: ListNode
		:rtype: TreeNode
		"""

		def getNodeNum(head):
			NodeNum = 0

			while head != None:
				head = head.next
				NodeNum += 1

			return NodeNum

		def getNodeBeforeMedium(head):
			NodeNum = getNodeNum(head)
			midPrevNode = head

			for i in xrange(NodeNum / 2 - 1):
				midPrevNode = midPrevNode.next

			return midPrevNode

		if head is None:
			return None

		if head.next is None:
			return TreeNode(head.val)

		midPrevNode = getNodeBeforeMedium(head)
		root = TreeNode(midPrevNode.next.val)
		root.right = self.sortedListToBST(midPrevNode.next.next)
		midPrevNode.next = None
		root.left = self.sortedListToBST(head)

		return root
